`HistogramViewer` and `ImageViewer`

kangz12345 opened the pull request:
This closes #130.

You can test this in `kangz12345/130/histogramviewer-and-imageviewer-demo`.

![Screen Shot 2023-09-14 at 16 17 41](https://github.com/snu-quiqcl/iquip/assets/8445906/6c9d65ed-6540-4c48-88a5-6c20ec6a3d35)

BECATRUE commented on iquip/apps/dataviewer.py:
```diff
@@ -92,3 +92,68 @@ def setData(self, data: np.ndarray, axes: Sequence[AxisInfo]):
+        super().__init__(ndim=1)
+        self.plotItem = pg.PlotItem(**kwargs)
+        self.widget = pg.PlotWidget(plotItem=self.plotItem)
+        self.histogram = pg.BarGraphItem(x=(), height=(), width=1)
```
Why did you pass an empty tuple explicitly? Isn't able to remove it?

BECATRUE replied:
I mean that `pg.BarGraphItem()` isn't okay?

kangz12345 replied:
They are necessary. If we don't give any parameters, it occurs an error.
You can take a look at the [source code](https://pyqtgraph.readthedocs.io/en/latest/_modules/pyqtgraph/graphicsItems/BarGraphItem.html).
To be more specific, `_getNormalizedCoords()` needs actual data.

BECATRUE replied:
I see. Then your implementation looks correct!

BECATRUE commented on iquip/apps/dataviewer.py:
```diff
@@ -92,3 +92,68 @@ def setData(self, data: np.ndarray, axes: Sequence[AxisInfo]):
+        self.plotItem.setLabel(axis="left", text=vaxis.name, units=vaxis.unit)
+        self.plotItem.setLabel(axis="bottom", text=haxis.name, units=haxis.unit)
+        x, y = haxis.values[0], vaxis.values[0]
+        width, height = haxis.values[-1] - x, vaxis.values[-1] - y
```
It forces the axes value should be a number.
Does the docstring in terms of "linearly increasing sequences" imply this information?

kangz12345 replied:
Yes, but it looks a bit ambiguous. I will update the docstring of `AxisInfo` since `pyqtgraph` seems only support numeric values 😅

kangz12345 commented:
![Screen Shot 2023-09-15 at 10 40 04](https://github.com/snu-quiqcl/iquip/assets/8445906/69996713-128e-4e66-ae13-e00e366816cf)

Hmm.. unit test aborts because of the thread issue I guess..?

kangz12345 commented:
> ![Screen Shot 2023-09-15 at 10 40 04](https://user-images.githubusercontent.com/8445906/268148729-69996713-128e-4e66-ae13-e00e366816cf.png)
> 
> Hmm.. unit test aborts because of the thread issue I guess..?

Never mind, I re-ran the actions and got the usual errors again.

kangz12345 commented:
> > Hmm.. unit test aborts because of the thread issue I guess..?
> 
> Never mind, I re-ran the actions and got the usual errors again.

Oh, actually now the thread issue comes up repeatedly.

kangz12345 commented:
I found that the unit test aborts even on my machine.

kangz12345 commented:
I will rebase this branch onto the current `develop` branch since it diverged before the #135 is merged.

kangz12345 committed:
Try to fix unit test action error

kangz12345 committed:
Unlock the aspect ratio for `ImageViewer`

kangz12345 committed:
Restrict axis values to be numeric only

kangz12345 committed:
Fix error due to `image=None`

kangz12345 committed:
Change attribute name `plotWidget` -> `widget`

kangz12345 committed:
Introduce `ImageViewer`

kangz12345 committed:
Transform the image scale and position

For showing proper axis values.

kangz12345 committed:
Add attribute docstring for `HistogramViewer`

kangz12345 committed:
Fix typo

kangz12345 committed:
Show grid in `ImageViewer` by default

kangz12345 committed:
Remove unused import

kangz12345 committed:
Implement `HistogramViewer`

Even if it has many things in common with `CurvePlotViewer`,
I will use it as now at the moment as it could be changed later.

kangz12345 committed:
Try to fix the unit test action error

kangz12345 committed:
Roll back the change in unittest.yml

BECATRUE commented:
Now, the unittest doesn't be aborted! (Sure, it still fails due to using global constants. I will resolve it as soon as possible.)

kangz12345 commented:
> Now, the unittest doesn't be aborted! (Sure, it still fails due to using global constants. I will resolve it as soon as possible.)

I think it is non-deterministic because it is due to threads.
